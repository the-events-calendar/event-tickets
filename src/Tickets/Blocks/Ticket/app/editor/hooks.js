import { store } from '@moderntribe/common/store';
import { actions } from '@moderntribe/tickets/data/blocks/ticket';
import { wpHooks } from '@moderntribe/common/utils/globals';

const filterBlockEdit = ( BlockEdit ) => ( props ) => {
	const { dispatch } = store;

	wp.element.useEffect( () => {
		const { name, clientId } = props;

		return () => {
			if ( name === 'tribe/tickets-item' ) {
				dispatch( actions.deleteTicket( clientId, false ) );
			}
		};
	}, [] );

	return wp.element.createElement(
		wp.element.Fragment,
		null,
		wp.element.createElement( BlockEdit, props ),
	);
};

/**
 * Filter to determine if a block was deleted using the delete block option,
 * also validates if its a ticket block, then call deleteTicket on unmount
 */
export const initHook = () => {
	wpHooks.addFilter(
		'editor.BlockEdit',
		'event-tickets',
		filterBlockEdit,
	);
};
