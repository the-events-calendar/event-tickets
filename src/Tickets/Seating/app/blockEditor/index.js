import './store';
import './filters.js';
