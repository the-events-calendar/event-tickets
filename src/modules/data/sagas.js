/**
 * Internal dependencies
 */
import { store } from '@moderntribe/common/store';
import { sagas as RSVPSagas } from './blocks/rsvp';
import { sagas as TicketSagas } from './blocks/ticket';
import { sagas as AttendeesSagas } from './blocks/attendees';
import MoveSagas from './shared/move/sagas';

export default () => {
	[ RSVPSagas, TicketSagas, AttendeesSagas, MoveSagas ].forEach( ( sagas ) => store.run( sagas ) );
};
