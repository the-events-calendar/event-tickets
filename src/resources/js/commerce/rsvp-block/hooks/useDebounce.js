/**
 * Debounce hook for React
 *
 * @since TBD
 */
import { useState, useEffect } from '@wordpress/element';

/**
 * Hook to debounce a value
 *
 * @since TBD
 *
 * @param {any} value The value to debounce.
 * @param {number} delay The delay in milliseconds.
 *
 * @return {any} The debounced value.
 */
export function useDebounce( value, delay ) {
	const [ debouncedValue, setDebouncedValue ] = useState( value );

	useEffect( () => {
		const handler = setTimeout( () => {
			setDebouncedValue( value );
		}, delay );

		return () => {
			clearTimeout( handler );
		};
	}, [ value, delay ] );

	return debouncedValue;
}