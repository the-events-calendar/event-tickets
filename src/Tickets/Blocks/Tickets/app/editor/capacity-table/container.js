/**
 * External dependencies
 */
import { connect } from 'react-redux';
import { compose } from 'redux';

/**
 * Internal dependencies
 */
import CapacityTable from './template';
import { withStore } from '@moderntribe/common/hoc';
import { selectors, actions } from '@moderntribe/tickets/data/blocks/ticket';
import { __ } from '@wordpress/i18n';
import { applyFilters } from '@wordpress/hooks';

const getTicketItems = (tickets) => {
	const items = tickets
		.filter((ticket) => ticket.details.title)
		.map((ticket) => ticket.details.title)
		.join(', ');
	return items ? ` (${items}) ` : '';
};

const getIndependentTicketItems = (state) => {
	const independentTickets = selectors.getIndependentTickets(state);
	return getTicketItems(independentTickets);
};

const getSharedTicketItems = (state) => {
	const sharedTickets = selectors.getSharedTickets(state);
	return getTicketItems(sharedTickets);
};

const mapStateToProps = (state) => {
	const independentAndSharedCapacity =
		selectors.getIndependentAndSharedTicketsCapacity(state);
	const unlimitedTicketItems = getTicketItems(
		selectors.getUnlimitedTickets(state)
	);
	const totalCapacity = unlimitedTicketItems.length
		? __('Unlimited', 'event-tickets')
		: independentAndSharedCapacity;

	let mappedProps = {
		independentCapacity: selectors.getIndependentTicketsCapacity(state),
		independentTicketItems: getIndependentTicketItems(state),
		isSettingsLoading: selectors.getTicketsIsSettingsLoading(state),
		rowsAfter: [],
		sharedCapacity: selectors.getTicketsSharedCapacity(state),
		sharedTicketItems: getSharedTicketItems(state),
		totalCapacity,
		unlimitedTicketItems,
	};

	/**
	 * Filters the properties mapped from the state for the CapacityTable component.
	 *
	 * @since 5.8.0
	 *
	 * @param {Object} mappedProps   The mapped props.
	 * @param {Object} context.state The state of the block.
	 */
	mappedProps = applyFilters(
		'tec.tickets.blocks.Tickets.CapacityTable.mappedProps',
		mappedProps,
		{ state }
	);

	return mappedProps;
};

const mapDispatchToProps = (dispatch) => ({
	onSharedCapacityChange: (e) => {
		dispatch(actions.setTicketsSharedCapacity(e.target.value));
		dispatch(actions.setTicketsTempSharedCapacity(e.target.value));
	},
});

export default compose(
	withStore(),
	connect(mapStateToProps, mapDispatchToProps)
)(CapacityTable);
