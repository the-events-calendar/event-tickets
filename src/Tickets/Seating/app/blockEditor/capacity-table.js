import { select } from '@wordpress/data';
import { storeName } from './store';
import { getLocalizedString } from '@tec/tickets/seating/utils';

/**
 * Filters the mapped props for the Capacity Table component.
 *
 *
 * @since 5.16.0
 *
 * @param {Object} mappedProps The mapped props for the Capacity Table component.
 *
 * @return {Object} The mapped props for the Capacity Table component.
 */
export function filterCapacityTableMappedProps(mappedProps) {
	const store = select(storeName);
	const hasSeats = store.isUsingAssignedSeating();
	const layoutLocked = store.isLayoutLocked();

	if (!(hasSeats && layoutLocked)) {
		return mappedProps;
	}

	const layoutId = store.getCurrentLayoutId();
	if (!layoutId) {
		return mappedProps;
	}

	const seatTypes = store.getSeatTypesForLayout(layoutId, true);
	const activeSeatTypes = Object.values(store.getSeatTypesByPostID()).filter(
		(value, index, array) => array.indexOf(value) === index
	);

	if (!Object.keys(seatTypes).length || !activeSeatTypes.length) {
		return mappedProps;
	}

	mappedProps.rowsAfter = mappedProps.rowsAfter || [];
	const seatTypeLabels = activeSeatTypes.map((type) => seatTypes[type].name);
	const seatTypeTotalCapacity = activeSeatTypes.reduce(
		(sum, type) => sum + parseInt(seatTypes[type].seats),
		0
	);
	mappedProps.rowsAfter.push({
		label: getLocalizedString('seats-row-label', 'capacity-table'),
		items: seatTypeLabels ? `(${seatTypeLabels})` : '',
		right: String(seatTypeTotalCapacity),
	});

	mappedProps.totalCapacity =
		mappedProps.totalCapacity -
		mappedProps.sharedCapacity +
		seatTypeTotalCapacity;
	mappedProps.sharedCapacity = '';
	mappedProps.sharedTicketItems = '';

	return mappedProps;
}
