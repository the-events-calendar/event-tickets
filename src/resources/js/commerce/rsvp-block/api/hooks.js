/**
 * React Query hooks for RSVP API operations
 *
 * @since TBD
 */
import { useQuery, useMutation, useQueryClient } from '@tanstack/react-query';
import { useSelect } from '@wordpress/data';
import { createRSVP, updateRSVP, deleteRSVP, fetchRSVP } from './endpoints';

/**
 * Hook for creating a new RSVP
 *
 * @since TBD
 *
 * @return {Object} Mutation object from React Query.
 */
export const useCreateRSVP = () => {
	const queryClient = useQueryClient();
	const postId = useSelect( ( select ) => select( 'core/editor' ).getCurrentPostId() );

	return useMutation( {
		mutationFn: ( data ) => createRSVP( { ...data, postId } ),
		onSuccess: ( data ) => {
			// Invalidate queries to refresh data
			queryClient.invalidateQueries( { queryKey: [ 'rsvp', postId ] } );
			
			// Set the new RSVP data in cache
			if ( data.ticket_id ) {
				queryClient.setQueryData( [ 'rsvp', String( data.ticket_id ) ], data );
			}
		},
		onError: ( error ) => {
			console.error( 'Failed to create RSVP:', error );
		},
	} );
};

/**
 * Hook for updating an existing RSVP
 *
 * @since TBD
 *
 * @return {Object} Mutation object from React Query.
 */
export const useUpdateRSVP = () => {
	const queryClient = useQueryClient();
	const postId = useSelect( ( select ) => select( 'core/editor' ).getCurrentPostId() );

	return useMutation( {
		mutationFn: ( data ) => updateRSVP( { ...data, postId } ),
		onSuccess: ( data, variables ) => {
			// Update the cached data
			if ( variables.rsvpId ) {
				queryClient.setQueryData( [ 'rsvp', variables.rsvpId ], data );
			}
			
			// Invalidate related queries
			queryClient.invalidateQueries( { queryKey: [ 'rsvp' ] } );
		},
		// Optimistic update for better UX
		onMutate: async ( newData ) => {
			// Cancel any outgoing refetches
			await queryClient.cancelQueries( { queryKey: [ 'rsvp', newData.rsvpId ] } );

			// Snapshot the previous value
			const previousData = queryClient.getQueryData( [ 'rsvp', newData.rsvpId ] );

			// Optimistically update to the new value
			if ( newData.rsvpId ) {
				queryClient.setQueryData( [ 'rsvp', newData.rsvpId ], ( old ) => ( {
					...old,
					...newData,
				} ) );
			}

			// Return a context with the previous data
			return { previousData };
		},
		// If the mutation fails, use the context to roll back
		onError: ( err, newData, context ) => {
			if ( context?.previousData && newData.rsvpId ) {
				queryClient.setQueryData( [ 'rsvp', newData.rsvpId ], context.previousData );
			}
		},
		// Always refetch after error or success
		onSettled: ( data, error, variables ) => {
			if ( variables?.rsvpId ) {
				queryClient.invalidateQueries( { queryKey: [ 'rsvp', variables.rsvpId ] } );
			}
		},
	} );
};

/**
 * Hook for deleting an RSVP
 *
 * @since TBD
 *
 * @return {Object} Mutation object from React Query.
 */
export const useDeleteRSVP = () => {
	const queryClient = useQueryClient();
	const postId = useSelect( ( select ) => select( 'core/editor' ).getCurrentPostId() );

	return useMutation( {
		mutationFn: ( data ) => deleteRSVP( { ...data, postId } ),
		onSuccess: ( data, variables ) => {
			// Remove from cache
			if ( variables.rsvpId ) {
				queryClient.removeQueries( { queryKey: [ 'rsvp', variables.rsvpId ] } );
			}
			
			// Invalidate post RSVPs
			queryClient.invalidateQueries( { queryKey: [ 'rsvp', postId ] } );
		},
		onError: ( error ) => {
			console.error( 'Failed to delete RSVP:', error );
		},
	} );
};

/**
 * Hook for fetching RSVP data
 *
 * @since TBD
 *
 * @param {string} rsvpId The RSVP ID to fetch.
 *
 * @return {Object} Query object from React Query.
 */
export const useRSVP = ( rsvpId ) => {
	return useQuery( {
		queryKey: [ 'rsvp', rsvpId ],
		queryFn: () => fetchRSVP( rsvpId ),
		enabled: !! rsvpId,
		staleTime: 5 * 60 * 1000, // 5 minutes
		cacheTime: 10 * 60 * 1000, // 10 minutes
	} );
};

/**
 * Hook for fetching all RSVPs for a post
 *
 * @since TBD
 *
 * @return {Object} Query object from React Query.
 */
export const usePostRSVPs = () => {
	const postId = useSelect( ( select ) => select( 'core/editor' ).getCurrentPostId() );

	return useQuery( {
		queryKey: [ 'rsvp', 'post', postId ],
		queryFn: async () => {
			if ( ! postId ) return [];
			
			try {
				// Fetch all tickets for this post via REST API
				const response = await fetch( `/wp-json/tribe/tickets/v1/tickets?include_post=${postId}` );
				
				if ( ! response.ok ) {
					console.error( 'Failed to fetch tickets:', response.statusText );
					return [];
				}
				
				const data = await response.json();
				const tickets = data?.tickets || [];
				
				// Filter for tc_rsvp type tickets
				const rsvps = tickets.filter( ticket => 
					ticket.type === 'tc_rsvp' || 
					ticket.provider_class === 'TEC\\Tickets\\Commerce\\Module'
				);
				
				return rsvps;
			} catch ( error ) {
				console.error( 'Error fetching RSVPs:', error );
				return [];
			}
		},
		enabled: !! postId,
		staleTime: 5 * 60 * 1000, // 5 minutes
		cacheTime: 10 * 60 * 1000, // 10 minutes
	} );
};