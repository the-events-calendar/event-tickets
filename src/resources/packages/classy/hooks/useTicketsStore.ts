import { useEffect } from 'react';
import { useSelect, useDispatch } from '@wordpress/data';
import { registerStore } from '@wordpress/data';
import { STORE_NAME } from '../store/constants';
import { storeConfig } from '../store';

// Register the store if it hasn't been registered yet
if ( !window.wp?.data?.select( STORE_NAME ) ) {
	registerStore( STORE_NAME, storeConfig );
}

export const useTicketsStore = () => {
	const dispatch = useDispatch();

	// Selectors
	const tickets = useSelect( ( select ) => {
		const store = select( STORE_NAME );
		return store ? store.getTicketsForPost( store.getState() ) : [];
	}, [] );

	const isLoading = useSelect( ( select ) => {
		const store = select( STORE_NAME );
		return store ? store.getIsLoading( store.getState() ) : false;
	}, [] );

	const error = useSelect( ( select ) => {
		const store = select( STORE_NAME );
		return store ? store.getError( store.getState() ) : null;
	}, [] );

	const currentPostId = useSelect( ( select ) => {
		const store = select( STORE_NAME );
		return store ? store.getCurrentPostId( store.getState() ) : null;
	}, [] );

	// Actions
	const fetchTickets = ( postId: number ) => {
		const store = dispatch( STORE_NAME );
		return store ? store.fetchTickets( postId ) : Promise.resolve();
	};

	const createTicket = ( ticketData: any ) => {
		const store = dispatch( STORE_NAME );
		return store ? store.createTicket( ticketData ) : Promise.resolve();
	};

	const updateTicket = ( ticketId: number, ticketData: any ) => {
		const store = dispatch( STORE_NAME );
		return store ? store.updateTicket( ticketId, ticketData ) : Promise.resolve();
	};

	const deleteTicket = ( ticketId: number ) => {
		const store = dispatch( STORE_NAME );
		return store ? store.deleteTicket( ticketId ) : Promise.resolve();
	};

	const clearError = () => {
		const store = dispatch( STORE_NAME );
		return store ? store.clearError() : undefined;
	};

	return {
		// State
		tickets,
		isLoading,
		error,
		currentPostId,

		// Actions
		fetchTickets,
		createTicket,
		updateTicket,
		deleteTicket,
		clearError,
	};
};
