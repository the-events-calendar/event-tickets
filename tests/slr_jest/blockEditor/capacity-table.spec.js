import { select } from '@wordpress/data';
import { applyFilters } from '@wordpress/hooks';
import { filterCapacityTableMappedProps } from '@tec/tickets/seating/blockEditor/capacity-table';

jest.mock('@wordpress/data', () => ({
	select: jest.fn(),
	createReduxStore: jest.fn(),
	register: jest.fn(),
}));

describe('filterCapacityTableMappedProps', () => {
	beforeEach(() => {
		jest.resetModules();
		jest.resetAllMocks();
	});

	afterEach(() => {
		jest.resetModules();
		jest.resetAllMocks();
	});

	it('returns original mappedProps if hasSeats is false', () => {
		select.mockReturnValue({
			isUsingAssignedSeating: () => false,
			isLayoutLocked: () => true,
		});
		const mappedProps = { foo: 'bar' };
		const newMappedProps = filterCapacityTableMappedProps(mappedProps);
		expect(newMappedProps).toEqual(mappedProps);
	});

	it('returns original mappedProps if layoutLocked is false', () => {
		select.mockReturnValue({
			isUsingAssignedSeating: () => true,
			isLayoutLocked: () => false,
		});
		const mappedProps = { foo: 'bar' };
		const newMappedProps = filterCapacityTableMappedProps(mappedProps);
		expect(newMappedProps).toEqual(mappedProps);
	});

	it('returns original mappedProps if layoutId is null', () => {
		select.mockReturnValue({
			isUsingAssignedSeating: () => true,
			isLayoutLocked: () => true,
			getCurrentLayoutId: () => null,
		});
		const mappedProps = { foo: 'bar' };
		const newMappedProps = filterCapacityTableMappedProps(mappedProps);
		expect(newMappedProps).toEqual(mappedProps);
	});

	it('adds rowsAfter and updates totalCapacity if hasSeats and layoutLocked are true', () => {
		const seatTypes = {
			vip: { name: 'VIP', seats: 20 },
			'general-admission': { name: 'General Admission', seats: 20 },
		};
		const activeSeatTypes = {
			99: 'vip',
		};
		select.mockReturnValue({
			isUsingAssignedSeating: () => true,
			isLayoutLocked: () => true,
			getCurrentLayoutId: () => 'layout-id',
			getSeatTypesForLayout: () => seatTypes,
			getSeatTypesByPostID: () => activeSeatTypes,
		});

		const mappedProps = {
			rowsAfter: [],
			independentCapacity: 20,
			totalCapacity: 40,
			sharedCapacity: 20,
			sharedTicketItems: 'items',
		};

		const result = filterCapacityTableMappedProps(mappedProps);
		expect(result).toMatchSnapshot();
	});

	it('adds rowsAfter and updates totalCapacity if has multiple seat types', () => {
		const seatTypes = {
			vip: { name: 'VIP', seats: 20 },
			'general-admission': { name: 'General Admission', seats: 20 },
		};
		const activeSeatTypes = {
			99: 'vip',
			100: 'general-admission',
		};
		select.mockReturnValue({
			isUsingAssignedSeating: () => true,
			isLayoutLocked: () => true,
			getCurrentLayoutId: () => 'layout-id',
			getSeatTypesForLayout: () => seatTypes,
			getSeatTypesByPostID: () => activeSeatTypes,
		});

		const mappedProps = {
			rowsAfter: [],
			independentCapacity: 20,
			totalCapacity: 60,
			sharedCapacity: 40,
			sharedTicketItems: 'items',
		};

		const result = filterCapacityTableMappedProps(mappedProps);
		expect(result).toMatchSnapshot();
	});

	it('adds rowsAfter and updates totalCapacity if has multiple seat types and multipe of the same type', () => {
		const seatTypes = {
			vip: { name: 'VIP', seats: 20 },
			'general-admission': { name: 'General Admission', seats: 20 },
			'ultra-vip': { name: 'Ultra VIP', seats: 40 },
		};
		const activeSeatTypes = {
			99: 'vip',
			100: 'vip',
			101: 'general-admission',
			102: 'general-admission',
			103: 'ultra-vip',
		};
		select.mockReturnValue({
			isUsingAssignedSeating: () => true,
			isLayoutLocked: () => true,
			getCurrentLayoutId: () => 'layout-id',
			getSeatTypesForLayout: () => seatTypes,
			getSeatTypesByPostID: () => activeSeatTypes,
		});

		const mappedProps = {
			rowsAfter: [],
			independentCapacity: 20,
			totalCapacity: 60,
			sharedCapacity: 40,
			sharedTicketItems: 'items',
		};

		const result = filterCapacityTableMappedProps(mappedProps);
		expect(result).toMatchSnapshot();
	});
});
