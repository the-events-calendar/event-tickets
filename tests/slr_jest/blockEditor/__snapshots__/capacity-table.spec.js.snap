// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filterCapacityTableMappedProps adds rowsAfter and updates totalCapacity if has multiple seat types 1`] = `
Object {
  "independentCapacity": 20,
  "rowsAfter": Array [
    Object {
      "items": "(VIP,General Admission)",
      "label": "Assigned seating",
      "right": "40",
    },
  ],
  "sharedCapacity": "",
  "sharedTicketItems": "",
  "totalCapacity": 60,
}
`;

exports[`filterCapacityTableMappedProps adds rowsAfter and updates totalCapacity if has multiple seat types and multipe of the same type 1`] = `
Object {
  "independentCapacity": 20,
  "rowsAfter": Array [
    Object {
      "items": "(VIP,General Admission,Ultra VIP)",
      "label": "Assigned seating",
      "right": "80",
    },
  ],
  "sharedCapacity": "",
  "sharedTicketItems": "",
  "totalCapacity": 100,
}
`;

exports[`filterCapacityTableMappedProps adds rowsAfter and updates totalCapacity if hasSeats and layoutLocked are true 1`] = `
Object {
  "independentCapacity": 20,
  "rowsAfter": Array [
    Object {
      "items": "(VIP)",
      "label": "Assigned seating",
      "right": "20",
    },
  ],
  "sharedCapacity": "",
  "sharedTicketItems": "",
  "totalCapacity": 40,
}
`;
